### PR TITLE
[expo-notifications] Move notification response handling to delegate

### PR DESCRIPTION
**Pull request by sjchmiela**
# Why

1. Notification response handling doesn't work after recent changes as per https://github.com/expo/expo/pull/10708#pullrequestreview-511839659.
2. Simplify implementation.
3. Single-threaded notifications handling

# How

Instead of relying on `NotificationResponseReceiver` which uses `NotificationsHelper` to dispatch events to `NotificationsService` we can handle them in `NotificationsService` in the first place. Basically I:
- removed `*NotificationResponseReceiver` by using _action intent_ created by new companion object method in `NotificationsService`
- changed `RECEIVE_RESPONSE` event from `(notificationResponse: NotificationResponse, opensAppToForegound: Boolean)` to `(action: NotificationAction, notification: Notification)`. This lets us create notification responses when we actually have them (previously we created text response, put it into the intent and after receiving text we modified the original response object). This also makes it easier to extend `NotificationAction` API by not adding more simple flags to intent data and depending on the actual action object itself.
- moved opening app to foreground to service's handling delegate which makes it overridable
- removed need for `CATEGORY_DEFAULT` from activities wanting to handle `expo.modules.notifications.OPEN_APP_ACTION` — previously we used `intent.resolveActivity` which adds `PackageManager.MATCH_DEFAULT_ONLY` flag to search, now we use `packageManager` without flags

less important:

- added information about `opensAppToForeground` defaulting to `true`
- added `${Constants.manifest.id}:welcome` category creation to NCL for the test push notification to test the category

# Test Plan

Verified that tapping the notification and different actions is handled properly by the experience.

I've spent much time debugging why notifications aren't properly scheduled when the device boots only to find out Nokia device I was using to test [has some DuraSpeed system](https://github.com/urbandroid-team/dont-kill-my-app/issues/57) that force stops all apps at boot and forbids them receive scheduled broadcasts. On emulator everything works perfect.

### CODE REVIEW

File: packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/NotificationsService.kt [hunk 1]
```diff
@@ -134,7 +142,7 @@ open class NotificationsService : BroadcastReceiver() {
     fun handleResponseReceived(context: Context, response: NotificationResponse, receiver: ResultReceiver? = null) {
       val data = getUriBuilderForIdentifier(response.notification.notificationRequest.identifier).appendPath("response").build()
       doWork(context, Intent(NOTIFICATION_EVENT_ACTION, data).also { intent ->
-        intent.putExtra(EVENT_TYPE_KEY, RECEIVE_TYPE)
+        intent.putExtra(EVENT_TYPE_KEY, RECEIVE_RESPONSE_TYPE)
         intent.putExtra(NOTIFICATION_RESPONSE_KEY, response)
```
Comment (cruzach) [inline packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/NotificationsService.kt:146]: can we remove this function now? It looks like this is called from `NotificationsHelper.responseReceived`, but all those calls were in the `NotificationResponseReceiver` and `TextInputNotificationResponseReceiver`

**Review by cruzach — COMMENTED** [review]
I much prefer this way of handling this all in `NotificationsService` 😄 

I'm able to get NotificationResponses coming through just fine in cases where the response is to an interactive notification (e.g. button or text response), but just tapping on the notification doesn't seem to be triggering any response 🤔  this is on a pixel, Android 10

**Comment by sjchmiela** [comment]
Hm… I'm getting responses no matter the type of the interaction, eg. when tapping on a <kbd>Present notification immediately</kbd> in NCL and then on the notification I got

```js
Object {
  "notificationResponse": Object {
    "actionIdentifier": "expo.modules.notifications.actions.DEFAULT",
    "notification": Object {
      "date": 1603476575385,
      "request": Object {
        "content": Object {
          "autoDismiss": true,
          "badge": null,
          "body": "This is the body",
          "data": Object {
            "future": "self",
            "hello": "there",
          },
          "sound": "default",
          "sticky": false,
          "subtitle": null,
          "title": "Here is a scheduled notification!",
        },
        "identifier": "a567944f-2416-4794-8a6c-3087b2172604",
        "trigger": null,
      },
    },
  },
}
```

and when tapping on a push notification (triggered by <kbd>Send me a push notification</kbd> button) I get

```js
Object {
  "notificationResponse": Object {
    "actionIdentifier": "expo.modules.notifications.actions.DEFAULT",
    "notification": Object {
      "date": 1603476645556,
      "request": Object {
        "content": Object {
          "autoDismiss": true,
          "badge": null,
          "body": "Native Component List is registered for push notifications.",
          "categoryIdentifier": "@sjchmiela/native-component-list:welcome",
          "data": Object {
            "example": "sample data",
          },
          "sound": "default",
          "sticky": false,
          "subtitle": null,
          "title": "Welcome to Expo!",
        },
        "identifier": "0:1603476645560988%0ac519e6f9fd7ecd",
        "trigger": Object {
          "channelId": null,
          "remoteMessage": Object {
            "collapseKey": null,
            "data": Object {
              "body": "{\"example\":\"sample data\"}",
              "categoryId": "@sjchmiela/native-component-list:welcome",
              "experienceId": "@sjchmiela/native-component-list",
              "message": "Native Component List is registered for push notifications.",
              "title": "Welcome to Expo!",
            },
            "from": "367315174693",
            "messageId": "0:1603476645560988%0ac519e6f9fd7ecd",
            "messageType": null,
            "notification": null,
            "originalPriority": 2,
            "priority": 2,
            "sentTime": 1603476645556,
            "to": null,
            "ttl": 2419200,
          },
          "type": "push",
        },
      },
    },
  },
}
```

**Comment by cruzach** [comment]
@sjchmiela you're right, I must have done something wrong (not sure what 🤷 ) but testing it again the listener seems to be firing as expected with any type of notification response. Sorry about that mixup on my part!

**Comment by sjchmiela** [comment]
No problem! Always glad to reensure the change works plus it's super reassuring to know I'm not the only one testing the changes! 😍

**Pull request merged**